File: backend/models/debug.py
"""
调试数据模型 — 脚本执行调试信息

用于存储和传输脚本执行过程中的调试信息，包括日志、变量快照、性能指标等。
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class LogLevel(str, Enum):
    """日志级别"""
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"


class DebugLog(BaseModel):
    """调试日志条目"""
    timestamp: datetime = Field(default_factory=datetime.now, description="日志时间戳")
    level: LogLevel = Field(default=LogLevel.INFO, description="日志级别")
    message: str = Field(..., description="日志消息")
    line_number: int | None = Field(None, description="代码行号")
    context: dict[str, Any] | None = Field(None, description="上下文信息")


class VariableSnapshot(BaseModel):
    """变量快照"""
    name: str = Field(..., description="变量名")
    value: Any = Field(..., description="变量值")
    type_name: str = Field(..., description="变量类型")
    size_bytes: int | None = Field(None, description="内存占用（字节）")


class PerformanceMetrics(BaseModel):
    """性能指标"""
    execution_time_ms: float = Field(..., description="执行时间（毫秒）")
    memory_peak_mb: float | None = Field(None, description="峰值内存（MB）")
    cpu_time_ms: float | None = Field(None, description="CPU 时间（毫秒）")
    lines_executed: int | None = Field(None, description="执行的代码行数")


class DebugExecutionResult(BaseModel):
    """调试执行结果"""
    execution_id: str = Field(..., description="执行 ID")
    success: bool = Field(..., description="是否成功")
    result: Any = Field(None, description="执行结果")
    error: str | None = Field(None, description="错误信息")

    # 调试信息
    logs: list[DebugLog] = Field(default_factory=list, description="执行日志")
    variables: list[VariableSnapshot] = Field(default_factory=list, description="变量快照")
    performance: PerformanceMetrics = Field(..., description="性能指标")

    # 元数据
    script_code: str = Field(..., description="执行的脚本代码")
    started_at: datetime = Field(default_factory=datetime.now, description="开始时间")
    completed_at: datetime | None = Field(None, description="完成时间")


class DebugRunRequest(BaseModel):
    """调试运行请求"""
    script_code: str = Field(..., description="脚本代码", min_length=1, max_length=50000)
    context: dict[str, Any] = Field(default_factory=dict, description="上下文变量")
    timeout: float = Field(default=5.0, description="超时时间（秒）", ge=0.1, le=30.0)
    capture_variables: bool = Field(default=True, description="是否捕获变量快照")
    capture_logs: bool = Field(default=True, description="是否捕获日志")


class DebugSession(BaseModel):
    """调试会话"""
    session_id: str = Field(..., description="会话 ID")
    created_at: datetime = Field(default_factory=datetime.now, description="创建时间")
    executions: list[DebugExecutionResult] = Field(default_factory=list, description="执行历史")

File: frontend/src/api/debug.ts
/**
 * 调试 API 封装
 *
 * 提供脚本调试相关的 API 调用方法
 */

import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8002'

export interface DebugRunRequest {
  script_code: string
  context?: Record<string, any>
  timeout?: number
  capture_variables?: boolean
  capture_logs?: boolean
}

export interface DebugLog {
  timestamp: string
  level: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR'
  message: string
  line_number?: number
  context?: Record<string, any>
}

export interface VariableSnapshot {
  name: string
  value: any
  type_name: string
  size_bytes?: number
}

export interface PerformanceMetrics {
  execution_time_ms: number
  memory_peak_mb?: number
  cpu_time_ms?: number
  lines_executed?: number
}

export interface DebugExecutionResult {
  execution_id: string
  success: boolean
  result: any
  error?: string
  logs: DebugLog[]
  variables: VariableSnapshot[]
  performance: PerformanceMetrics
  script_code: string
  started_at: string
  completed_at?: string
}

export interface DebugRunResponse {
  status: string
  data: DebugExecutionResult
}

export interface DebugHistoryItem {
  execution_id: string
  success: boolean
  started_at: string
  execution_time_ms: number
  error?: string
}

/**
 * 在调试模式下运行脚本
 */
export async function debugRunScript(request: DebugRunRequest): Promise<DebugRunResponse> {
  const response = await axios.post(`${API_BASE_URL}/rule-chain/debug/run`, request)
  return response.data
}

/**
 * 获取调试日志
 */
export async function getDebugLogs(executionId: string): Promise<DebugLog[]> {
  const response = await axios.get(`${API_BASE_URL}/rule-chain/debug/logs/${executionId}`)
  return response.data.logs
}

/**
 * 获取变量快照
 */
export async function getDebugVariables(executionId: string): Promise<VariableSnapshot[]> {
  const response = await axios.get(`${API_BASE_URL}/rule-chain/debug/variables/${executionId}`)
  return response.data.variables
}

/**
 * 获取调试历史
 */
export async function getDebugHistory(sessionId: string = 'default', limit: number = 10): Promise<DebugHistoryItem[]> {
  const response = await axios.get(`${API_BASE_URL}/rule-chain/debug/history`, {
    params: { session_id: sessionId, limit }
  })
  return response.data.history
}

/**
 * 清空调试历史
 */
export async function clearDebugHistory(sessionId: string = 'default'): Promise<void> {
  await axios.delete(`${API_BASE_URL}/rule-chain/debug/history/${sessionId}`)
}
